### PR TITLE
[Merged by Bors] - feat(Topology): Introduce HomeomorphClass

### PR DESCRIPTION
**Pull request by Thmoas-Guan**
Introduce HomeomorphClass

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Topology/Homeomorph.lean
Comment (j-loreaux) [inline Mathlib/Topology/Homeomorph.lean:1052]:
```suggestion
/-- `HomeomorphClass F A B` states that `F` is a type of homeomorphisms. -/
```

File: Mathlib/Topology/Homeomorph.lean
Comment (j-loreaux) [inline Mathlib/Topology/Homeomorph.lean:1063]: You need to give it the `@[coe]` attribute so that it will actually show up in the infoview as an uparrow. This is also necessary for interaction with `norm_cast`.
```suggestion
/-- Turn an element of a type `F` satisfying `HomeomorphClass F α β` into an actual
`Homeomorph`. This is declared as the default coercion from `F` to `α ≃ₜ β`. -/
@[coe]
```

File: Mathlib/Topology/Homeomorph.lean
Comment (j-loreaux) [inline Mathlib/Topology/Homeomorph.lean:1068]: In case you haven't seen it yet, you may want to see the [style guide](https://leanprover-community.github.io/contribute/style.html)
```suggestion
def toHomeomorph [h : HomeomorphClass F α β] (f : F) : α ≃ₜ β :=
  { (f : α ≃ β) with
    continuous_toFun := h.continuous_toFun f
    continuous_invFun := h.continuous_invFun f }
```

File: Mathlib/Topology/Homeomorph.lean
Comment (j-loreaux) [inline Mathlib/Topology/Homeomorph.lean:1071]: please ask on Zulip what is the correct class to use here. `CoeTC` is not preferred (although it's littered throughout Mathlib). We need to come to a consensus on Zulip about which `Coe` class is best for coercions from morphism classes to morphisms.

File: Mathlib/Topology/Homeomorph.lean
Comment (j-loreaux) [inline Mathlib/Topology/Homeomorph.lean:1074]: I don't think we have this for other morphism classes. I suspect it's not necessary as we can just apply the coercion to functions and then use `DFunLike.coe_injective` (probably after `simp`)

**Review by j-loreaux — COMMENTED** [review]
Thanks for this addition! Can you please also add the analogue of this lemma, and address my comments below?
```lean
@[simp]
theorem RingHom.coe_coe {α : Type u_2} {β : Type u_3} :
∀ {x : NonAssocSemiring α} {x_1 : NonAssocSemiring β} {F : Type u_5} [inst : FunLike F α β]
  [inst_1 : RingHomClass F α β] (f : F), ⇑↑f = ⇑f
```

File: Mathlib/Topology/Homeomorph.lean
Comment (Thmoas-Guan) [reply Mathlib/Topology/Homeomorph.lean:1074]: I copied this lemma from `MulEquivClass`, I didn't think carefully whether it is useful.

File: Mathlib/Topology/Homeomorph.lean
Comment (Thmoas-Guan) [reply Mathlib/Topology/Homeomorph.lean:1071]: Asked in a thread of PR review, needed anywhere else?

Comment (Vierkantor) [reply Mathlib/Topology/Homeomorph.lean:1074]: The correct choice if you want to infer the left argument from the right is `CoeOut`. So:
```suggestion
instance [HomeomorphClass F α β] : CoeOut F (α ≃ₜ β) :=
  ⟨HomeomorphClass.toHomeomorph⟩
```

**Comment by Thmoas-Guan** [comment]
> ContinuousMapClass

So where should I add this, I have no idea.

**Comment by j-loreaux** [comment]
I think it's fine to import `Topology.ContinuousMap.Defs` here. This file is pretty large and should probably be split at some point in the future anyway.

File: Mathlib/Topology/Homeomorph.lean
Comment (j-loreaux) [reply Mathlib/Topology/Homeomorph.lean:1077]: Okay, it's fine to leave it then; perhaps it's useful. We don't have this lemma for most other coercions to morphisms though (which is why I didn't think it was likely to be useful).

**Comment by Thmoas-Guan** [comment]
> I think it's fine to import `Topology.ContinuousMap.Defs` here. It's lightweight and this file is pretty large and should probably be split at some point in the future anyway.

I think this PR is relative small, and I don't think leave it there and just let the mess grow is a very good idea, what about I just split `Homeomorph.lean` in this PR or I start a new one about it right now and resolve it once and for all. This would probably also unwind the import graph a bit.

**Comment by j-loreaux** [comment]
If you want to make a new PR splitting this file, go for it. I just didn't want to force you to do it.

I would suggest doing this in stages. First, create a PR which moves this to `Mathlib/Topology/Homeomorph/Basic.lean`. Then create a second PR which splits that file into (at least) a `Defs` file and maybe some others.

**Comment by Thmoas-Guan** [comment]
I just add this import for now, I'll deal the splitting latter.

**Comment by j-loreaux** [comment]
No, that would break things.

**Review by j-loreaux — COMMENTED** [review]
Can you please add instances of this class downstream (e.g., for `ContinuousSemilinearEquivClass`, and anything else that makes sense)?

**Comment by Thmoas-Guan** [comment]
Sorry that I am not familiar with these things and doesn't know what things look like downsteam, beside, is  `ContinuousSemilinearEquivClass` you mentioned also mixing algebra and topology and should be reconstructed? If so, that would be a big work and I probably wouldn't be able to work on it soon.

**Comment by j-loreaux** [comment]
Don't worry about refactoring `ContinuousSemilinearEquivClass`. All I'm asking for is an instance `ContinuousSemilinearEquivClass` → `HomeomorphClass`.

**Comment by Thmoas-Guan** [comment]
Added for now, but I am asking for the refactoring of it.

File: Mathlib/Topology/Algebra/Module/Basic.lean
Comment (kbuzzard) [inline Mathlib/Topology/Algebra/Module/Basic.lean:312]:
```suggestion
instance (priority := 100) homeomorphClass [EquivLike F M M₂]
```
I don't know this part of the library at all but if you're naming an instance manually it should surely start with a small letter (or maybe even `instHomeomorphClass`?)

Comment (Thmoas-Guan) [reply Mathlib/Topology/Algebra/Module/Basic.lean]: I just deleted the name and use the generated one. Thank you.

**Review by j-loreaux — APPROVED** [review]
Thanks!

maintainer merge

**Review by Vierkantor — APPROVED** [review]
Coming back to this to confirm that this looks good, `FunLike`-wise. I'd like to ask a topology maintainer to make the final decision.

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**
